### ISSUE TITLE
Correct metrics for gpu usage

### ISSUE THREAD

**Issue by DZADSL72-00558** [NVIDIA/TensorRT#2882]
## Description
Hello team, I'm a new GPU user. I was looking for the best statistic to use to gauge GPU use. The basic goal is to make use of all GPU processing resources.

The 'vidia-smi` reference I've made so far is very potent. However, the definition of GPU utilization from this library is as follows: 

```
Percent of time over the past sample period during which one or more kernels was executing on the GPU.
```

which was unexpected. I also attempted to examine the `pstate`,  which is similarly not very useful.

Another issue with `nvidia-smi` library is that I am unable to adjust with the sampling frequency, which is around 1/5 second. 

So, could you kindly teach me the proper measures and tools to determine whether I am making the most of my GPU? 

I'm investigating how the Nsight system can use SM utilization right now. Is this heading in the correct way? 

Any suggestions from you will be valued. 

## Environment

**TensorRT Version**: 
**NVIDIA GPU**: 
**NVIDIA Driver Version**: 
**CUDA Version**: 
**CUDNN Version**: 
**Operating System**: 
**Python Version (if applicable)**: 
**Tensorflow Version (if applicable)**: 
**PyTorch Version (if applicable)**: 
**Baremetal or Container (if so, version)**:

**Comment by Slyne** [comment]
Hi @DZADSL72-00558,
Not sure if my understanding is correctly. The measure metric you may want to check is [occupancy](https://docs.nvidia.com/gameworks/content/developertools/desktop/analysis/report/cudaexperiments/kernellevel/achievedoccupancy.htm). Using this nsight system to check sm utilization should be good.
However, this metric is limited to many factors, such as the underlying implementation of the cuda kernel.

Could you share more info about your application scenario?
If you are using data center, you may checkout DCGM.

**Comment by DZADSL72-00558** [comment]
Thanks, @Slyne! Is there a method for me to examine the TensorRT's underlying cuda kernels?  like each kernel's block size?

**Comment by Slyne** [comment]
> Thanks, @Slyne! Is there a method for me to examine the TensorRT's underlying cuda kernels? like each kernel's block size?

You probably can refer to [trtexec](https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/index.html#trtexec) for more info.
```
trtexec --dumpProfile --profilingVerbosity=detailed --dumpLayerInfo --verbose xxxx
```
BTW, tensorrt will choose the optimal configurations when generating the engine. If you have a performance issue, you may check [this](https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/index.html#report-performance-issue).

**Comment by DZADSL72-00558** [comment]
Hi @Slyne . Thanks for your suggestions. Actually, I tried [Nsight compute](https://docs.nvidia.com/nsight-compute/NsightCompute/index.html#occupancy-calculator).I received this from the occupancy calculator. Do you believe that this accurately depicts the proportion of GPU usage?  
![image (3)](https://user-images.githubusercontent.com/69881549/232928495-544de0f9-373e-4cc4-bba6-1011017d75a1.png)

**Comment by DZADSL72-00558** [comment]
I spent a while to read [this ](https://docs.nvidia.com/gameworks/content/developertools/desktop/analysis/report/cudaexperiments/kernellevel/achievedoccupancy.htm).It seems to be attempting to educate me how to optimize the amount of threads, registers, and shared memory per thread in order to get the greatest occupancy. It is very useful if I am a cuda developer. 

However, it is not what I'm searching for, because as you said, Tensorrt will choose the optimal configurations.

**Comment by Slyne** [comment]
Do you mean something like [GPU Metrics](https://docs.nvidia.com/nsight-systems/UserGuide/#gpu-metric-sampling) ?
Please add [this option](https://docs.nvidia.com/nsight-systems/UserGuide/#gms-from-CLI) when running your application.

Another straightforward way to check is simply:
![image](https://user-images.githubusercontent.com/6286804/233223113-7f360d8a-abbc-453b-ad79-9f23fb565ff7.png)

It looks like you are trying to optmize your gpu application. If you have a tensorrt engine, I would recommend to try some simple tricks:

1. Set fp16 option    fp16 usually runs faster than fp32. You can also let tensorrt to decide whether to fall back to fp32 or not by using --best
2. Set a larger workspace.
3. Create the network layer by layer with some existing plugins such as [here](https://github.com/NVIDIA/TensorRT/tree/master/plugin).  You may find this [onnx-graphsurgeon](https://github.com/NVIDIA/TensorRT/tree/master/tools/onnx-graphsurgeon) is also very useful.
4. If your neural network is very similar to transformer, you may try to check [FasterTransormer](https://github.com/NVIDIA/FasterTransformer/tree/main/examples) project.
5. Try int8  (Sometimes this will sacrafice accuracy. You may give it a try by inserting QDQ nodes and do some finetuning.)

The above are just my suggestions. Tensorrt developers are free to share any thoughts about them.

**Comment by DZADSL72-00558** [comment]
HI @Slyne .

In Nsight, I am able to locate the SM usage VS timeline. I'm grateful for your advice. 

Feel free to close this tt.

**Comment by DZADSL72-00558** [comment]
HI @Slyne, appreciate your help on it.

One final query regarding A10G. Do you happen to have the A10G white paper?  I looked for it on the internet but came up empty-handed.

**Comment by DZADSL72-00558** [comment]
Sorry for the noise, close it now.
